Source/NameServer/AddressBook.cpp

```diff
@@ -0,0 +1,1361 @@
+//==============================================================================
+#include "AddressBook.hpp"
+//#include "Debug.h"
+
+#include <iostream>
+
+//==============================================================================
+namespace AddressBookNS
+{
+#ifdef EXCEPTS
+#define ERETURN(X, Y) throw(X)
+#else
+#define ERETURN(X, Y) return Y
+#endif
+
+// static constant definitions
+const unsigned AddressBook::SUCCESS;
+const unsigned AddressBook::ERR_NONFATAL;
+const unsigned AddressBook::ERR_INVALID_TASK;
+const unsigned AddressBook::ERR_INVALID_DEVTYPE;
+const unsigned AddressBook::ERR_INVALID_DEVICE;
+const unsigned AddressBook::ERR_INVALID_MESSAGE_TYPE;
+const unsigned AddressBook::ERR_INVALID_ATTRIBUTE;
+const unsigned AddressBook::ERR_DEVICE_DATA_MISMATCH;
+const unsigned AddressBook::ERR_TASKNAME_USED;
+const unsigned AddressBook::ERR_DEVICENAME_USED;
+const unsigned AddressBook::ERR_DEVICE_ADDR_USED;
+const unsigned AddressBook::ERR_TASK_NOT_FOUND;
+const unsigned AddressBook::ERR_DEVICE_NOT_FOUND;
+const unsigned AddressBook::ERR_INVALID_MAP;
+const unsigned AddressBook::ERR_INVALID_SUPERVISOR;
+const unsigned AddressBook::ERR_INVALID_STATE;
+
+//Constructors
+AddressBook::AddressBook(std::string d)
+{
+   ABderived = d;
+   TaskCount = 0;
+}
+
+AddressBook::~AddressBook()
+{
+    // Teardown the TaskMap. Everything else should be handled properly.
+    for(TaskMap_t::iterator T=TaskMap.begin();T!=TaskMap.end();T++)
+    {
+        TaskRecord_t* TRec = &(*(T->second));
+        delete TRec;
+
+    }
+}
+
+
+/*==============================================================================
+ * GetTaskCount:    Get the number of currently loaded tasks
+ *============================================================================*/
+unsigned AddressBook::GetTaskCount(void)
+{
+    return TaskMap.size();
+}
+
+//==============================================================================
+//Task Manipulation
+unsigned AddressBook::AddTask(const std::string &TaskName, TaskData_t &Data)
+{
+    // Check that the task does not already exist.
```

Comment:
FWIW, this kind of repetition is fine because the actual check is abstracted into a method - the duplication is as small as it reasonably can be.